documentation/writing_tests.md

```diff
@@ -14,6 +20,139 @@
+
+Keep in mind that, over time, we might continue to expand these lists.
+
+Now, let's see how we're using the mentioned tools.
+
+### Mocha
+
+Mocha is a well known JavaScript test framework that runs in NodeJS and in the browser. If you're new to Mocha, we recommend you to dive into [Mocha's documentation](https://mochajs.org/).
+
+In this section, we'll explore how the Azure SDK for JS is using mocha, what versions we're using,how we're configuring it and any other recommendation on how to used this library.
+
+#### Mocha in our dependencies
+
+Since we're using [Rush](https://rushjs.io/), we're forced to use the same version of our packages in each one of the projects inside of this repository. If you want to add `mocha` as a dev dependency to a new project inside of this repository, first make sure you are in the root folder of that project, then you can run the following command:
+
+```
+rush add --dev -p mocha
+```
+
+To enhance the `mocha` experience, we're using the following packages:
+
+- [`@types/mocha`](https://www.npmjs.com/package/@types/mocha), the type definitions for mocha.
+- [`mocha-multi`](https://www.npmjs.com/package/mocha-multi), a way to get multiple reporters working with mocha.
+- [`mocha-junit-reporter`](https://www.npmjs.com/package/mocha-junit-reporter), which produces JUnit-style XML test results.
+- [`nyc`](https://www.npmjs.com/package/nyc) is [Istanbul](https://istanbul.js.org/)'s command line interface.
+- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
+- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://github.com/v8/v8/wiki/Stack-Trace-API).
+
+A full `rush add` command that includes `mocha` and all of the previous dependencies follows:
+
+```
+rush add --dev -p mocha @types/mocha mocha-multi mocha-junit-reporter nyc esm source-map-support
+```
+
+We'll explore how we use these dependencies up next.
+
+#### Configuring Mocha
+
+Mocha is used by our test scripts in our `package.json` files. We typically invoke it with some common parameters, and some parameters that help us specify different test targets and different ways to debug the tests.
+
+If we only consider the common parameters sent to mocha, we can see the following:
+
+```
+mocha --require esm --require source-map-support/register --reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=- --timeout 180000 --full-trace
+```
+
+Let's understand what's going on:
+
+- `--require esm` requires the `esm` ECMAScript module loader.
+- `--require source-map-support/register` requires the `source-map-support/register`, to support [source map](https://github.com/mozilla/source-map) for stack traces in node.
```

Comment:
interestingly, not all projects use this option (and it's not in the template), so we should fix that soon

Reply:
How can we fix this? I am learning about our dependencies as I'm writing this.